File: config-editor/src/lib/validation.ts
import type { MidiCaptainConfig } from './types';

export interface ValidationResult {
  isValid: boolean;
  errors: Map<string, string>;
}

export interface FieldValidator {
  (value: any, config?: MidiCaptainConfig): string | null;
}

export const validators = {
  label: (value: string): string | null => {
    if (!value || value.trim() === '') {
      return 'Label is required';
    }
    if (value.length > 6) {
      return 'Label must be 6 characters or less';
    }
    if (!/^[\w\s-]+$/.test(value)) {
      return 'Label contains invalid characters';
    }
    return null;
  },

  cc: (value: number, config?: MidiCaptainConfig): string | null => {
    if (value < 0 || value > 127) {
      return 'CC must be between 0 and 127';
    }
    if (!Number.isInteger(value)) {
      return 'CC must be an integer';
    }
    return null;
  },

  range: (min: number, max: number): string | null => {
    if (min >= max) {
      return 'Min must be less than max';
    }
    return null;
  },

  withinRange: (value: number, min: number, max: number): string | null => {
    if (value < min || value > max) {
      return `Value must be between ${min} and ${max}`;
    }
    return null;
  },

  note: (value: number): string | null => {
    if (value < 0 || value > 127) return 'Note must be between 0 and 127';
    if (!Number.isInteger(value)) return 'Note must be an integer';
    return null;
  },

  velocity: (value: number): string | null => {
    if (value < 0 || value > 127) return 'Velocity must be between 0 and 127';
    if (!Number.isInteger(value)) return 'Velocity must be an integer';
    return null;
  },

  program: (value: number): string | null => {
    if (value < 0 || value > 127) return 'Program must be between 0 and 127';
    if (!Number.isInteger(value)) return 'Program must be an integer';
    return null;
  },

  pcStep: (value: number): string | null => {
    if (!Number.isInteger(value)) return 'Step must be an integer';
    if (value < 1 || value > 127) return 'Step must be between 1 and 127';
    return null;
  },

  keytimes: (value: number): string | null => {
    if (!Number.isInteger(value)) return 'Keytimes must be an integer';
    if (value < 1 || value > 99) return 'Keytimes must be between 1 and 99';
    return null;
  },

  flashMs: (value: number): string | null => {
    if (!Number.isInteger(value)) return 'Flash duration must be an integer';
    if (value < 50 || value > 5000) return 'Flash duration must be between 50 and 5000 ms';
    return null;
  },

  // value is stored as 0-15 (displayed as 1-16)
  channel: (value: number): string | null => {
    if (!Number.isInteger(value)) return 'Channel must be an integer';
    if (value < 0 || value > 15) return 'Channel must be between 1 and 16';
    return null;
  },
};

/**
 * Validate an array of buttons (helper for single-bank and multi-bank configs)
 * @param buttons - Array of button configurations
 * @param pathPrefix - Path prefix for error keys (e.g., "buttons" or "banks[0].buttons")
 * @param errors - Map to accumulate errors
 */
function validateButtons(
  buttons: any[],
  pathPrefix: string,
  errors: Map<string, string>
): void {
  buttons.forEach((btn, idx) => {
    const btnPath = `${pathPrefix}[${idx}]`;
    
    const labelError = validators.label(btn.label);
    if (labelError) errors.set(`${btnPath}.label`, labelError);

    const msgType = btn.type ?? 'cc';

    if (msgType === 'cc') {
      if (btn.cc !== undefined) {
        const ccError = validators.cc(btn.cc);
        if (ccError) errors.set(`${btnPath}.cc`, ccError);
      }
    } else if (msgType === 'note') {
      if (btn.note !== undefined) {
        const noteError = validators.note(btn.note);
        if (noteError) errors.set(`${btnPath}.note`, noteError);
      }
      if (btn.velocity_on !== undefined) {
        const velError = validators.velocity(btn.velocity_on);
        if (velError) errors.set(`${btnPath}.velocity_on`, velError);
      }
      if (btn.velocity_off !== undefined) {
        const velError = validators.velocity(btn.velocity_off);
        if (velError) errors.set(`${btnPath}.velocity_off`, velError);
      }
    } else if (msgType === 'pc') {
      if (btn.program !== undefined) {
        const progError = validators.program(btn.program);
        if (progError) errors.set(`${btnPath}.program`, progError);
      }
    } else if (msgType === 'pc_inc' || msgType === 'pc_dec') {
      if (btn.pc_step !== undefined) {
        const stepError = validators.pcStep(btn.pc_step);
        if (stepError) errors.set(`${btnPath}.pc_step`, stepError);
      }
    }

    if (btn.channel !== undefined) {
      const chError = validators.channel(btn.channel);
      if (chError) errors.set(`${btnPath}.channel`, chError);
    }

    if (btn.flash_ms !== undefined) {
      const fError = validators.flashMs(btn.flash_ms);
      if (fError) errors.set(`${btnPath}.flash_ms`, fError);
    }

    // LED tap mode validation
    if (btn.led_mode !== undefined) {
      if (btn.led_mode !== 'tap') {
        errors.set(`${btnPath}.led_mode`, 'led_mode must be "tap" if set');
      }
      if (btn.tap_rate_ms !== undefined) {
        const tError = validators.flashMs(btn.tap_rate_ms);
        if (tError) errors.set(`${btnPath}.tap_rate_ms`, tError);
      }
    }

    // Long-press / long-release actions validation
    const validateAction = (action: any, pathBase: string) => {
      if (!action) return;
      if (typeof action !== 'object') {
        errors.set(pathBase, 'Action must be an object');
        return;
      }
      const aType = action.type ?? 'cc';
      if (!['cc', 'note', 'pc'].includes(aType)) {
        errors.set(`${pathBase}.type`, 'Action type must be cc, note, or pc');
      }
      if (aType === 'cc') {
        if (action.cc !== undefined) {
          const e = validators.cc(action.cc);
          if (e) errors.set(`${pathBase}.cc`, e);
        }
        if (action.value !== undefined) {
          const e = validators.withinRange(action.value, 0, 127);
          if (e) errors.set(`${pathBase}.value`, e);
        }
      } else if (aType === 'note') {
        if (action.note !== undefined) {
          const e = validators.note(action.note);
          if (e) errors.set(`${pathBase}.note`, e);
        }
        if (action.value !== undefined) {
          const e = validators.velocity(action.value);
          if (e) errors.set(`${pathBase}.value`, e);
        }
      } else if (aType === 'pc') {
        if (action.program !== undefined) {
          const e = validators.program(action.program);
          if (e) errors.set(`${pathBase}.program`, e);
        }
      }
      if (action.channel !== undefined) {
        const ch = validators.channel(action.channel);
        if (ch) errors.set(`${pathBase}.channel`, ch);
      }
      if (action.threshold_ms !== undefined) {
        if (!Number.isInteger(action.threshold_ms) || action.threshold_ms < 50 || action.threshold_ms > 10000) {
          errors.set(`${pathBase}.threshold_ms`, 'threshold_ms must be integer 50-10000');
        }
      }
    };

    // Validate long_press and long_release arrays
    const validateActionArray = (actions: any, pathBase: string) => {
      if (!actions) return;
      if (Array.isArray(actions)) {
        actions.forEach((action, idx) => {
          validateAction(action, `${pathBase}[${idx}]`);
          // threshold_ms only valid on first command
          if (idx > 0 && action.threshold_ms !== undefined) {
            errors.set(`${pathBase}[${idx}].threshold_ms`, 'threshold_ms only allowed on first command');
          }
        });
      } else {
        // Backward compatibility: treat single object as array of one
        validateAction(actions, pathBase);
      }
    };

    validateActionArray(btn.long_press, `${btnPath}.long_press`);
    validateActionArray(btn.long_release, `${btnPath}.long_release`);

    // select_group validation
    if (btn.select_group !== undefined) {
      if (typeof btn.select_group !== 'string' || btn.select_group.trim() === '') {
        errors.set(`${btnPath}.select_group`, 'select_group must be a non-empty string');
      } else {
        if (btn.mode === 'momentary') {
          errors.set(`${btnPath}.select_group`, 'select_group not supported for momentary mode');
        }
        if (btn.mode === 'tap') {
          errors.set(`${btnPath}.select_group`, 'select_group not supported for tap mode');
        }
        if ((btn.keytimes ?? 1) > 1) {
          errors.set(`${btnPath}.select_group`, 'select_group not supported with keytimes > 1');
        }
      }
    }
    if (btn.default_selected !== undefined && typeof btn.default_selected !== 'boolean') {
      errors.set(`${btnPath}.default_selected`, 'default_selected must be boolean');
    }

    if (btn.states && (btn.keytimes === undefined || btn.keytimes <= 1)) {
      errors.set(`${btnPath}.states`, 'states requires keytimes > 1');
    }

    if (btn.keytimes !== undefined) {
      const ktError = validators.keytimes(btn.keytimes);
      if (ktError) errors.set(`${btnPath}.keytimes`, ktError);

      if (btn.states) {
        btn.states.forEach((state: any, si: number) => {
          const sp = `${btnPath}.states[${si}]`;
          if (state.cc !== undefined) {
            const e = validators.cc(state.cc);
            if (e) errors.set(`${sp}.cc`, e);
          }
          if (state.cc_on !== undefined) {
            const e = validators.withinRange(state.cc_on, 0, 127);
            if (e) errors.set(`${sp}.cc_on`, e);
          }
          if (state.cc_off !== undefined) {
            const e = validators.withinRange(state.cc_off, 0, 127);
            if (e) errors.set(`${sp}.cc_off`, e);
          }
          if (state.note !== undefined) {
            const e = validators.note(state.note);
            if (e) errors.set(`${sp}.note`, e);
          }
          if (state.velocity_on !== undefined) {
            const e = validators.velocity(state.velocity_on);
            if (e) errors.set(`${sp}.velocity_on`, e);
          }
          if (state.velocity_off !== undefined) {
            const e = validators.velocity(state.velocity_off);
            if (e) errors.set(`${sp}.velocity_off`, e);
          }
          if (state.program !== undefined) {
            const e = validators.program(state.program);
            if (e) errors.set(`${sp}.program`, e);
          }
          if (state.pc_step !== undefined) {
            const e = validators.pcStep(state.pc_step);
            if (e) errors.set(`${sp}.pc_step`, e);
          }
          if (state.label !== undefined) {
            const e = validators.label(state.label);
            if (e) errors.set(`${sp}.label`, e);
          }
        });
      }
    }
  });

  // Cross-button select_group validation within this button set
  const groupDefaults: Record<string, number[]> = {};
  buttons.forEach((btn, idx) => {
    if (btn.select_group && btn.default_selected) {
      if (!groupDefaults[btn.select_group]) groupDefaults[btn.select_group] = [];
      groupDefaults[btn.select_group].push(idx);
    }
  });
  for (const g in groupDefaults) {
    const idxs = groupDefaults[g];
    if (idxs.length > 1) {
      for (let i = 1; i < idxs.length; i++) {
        errors.set(`${pathPrefix}[${idxs[i]}].default_selected`, `Multiple default_selected in group '${g}'`);
      }
    }
  }
}

export function validateConfig(config: MidiCaptainConfig): ValidationResult {
  const errors = new Map<string, string>();

  // Determine expected button count based on device
  const expectedButtons = config.device === 'mini6' ? 6 : 10;

  // Validate banks (multi-bank mode) or legacy buttons (single-bank mode)
  if (config.banks && config.banks.length > 0) {
    // Multi-bank mode validation
    if (config.banks.length > 8) {
      errors.set('banks', `Max 8 banks supported, found ${config.banks.length}`);
    }

    config.banks.forEach((bank, bankIdx) => {
      const bankPrefix = `banks[${bankIdx}]`;
      
      // Validate bank name
      if (!bank.name || bank.name.length === 0) {
        errors.set(`${bankPrefix}.name`, 'Bank name cannot be empty');
      } else if (bank.name.length > 20) {
        errors.set(`${bankPrefix}.name`, 'Bank name cannot exceed 20 chars');
      }

      // Validate button count
      if (!Array.isArray(bank.buttons)) {
        errors.set(`${bankPrefix}.buttons`, 'Bank buttons must be an array');
        return; // Skip further validation for this bank
      }
      if (bank.buttons.length !== expectedButtons) {
        errors.set(`${bankPrefix}.buttons`, `Expected ${expectedButtons} buttons for ${config.device}, found ${bank.buttons.length}`);
      }

      // Validate buttons for this bank
      validateButtons(bank.buttons, `${bankPrefix}.buttons`, errors);
    });

      // Validate each button in this bank
      validateButtons(bank.buttons, `${bankPrefix}.buttons`, errors);
    });

    // Validate bank_switch config if present
    if (config.bank_switch) {
      const bs = config.bank_switch;
      if (bs.channel !== undefined) {
        const chErr = validators.channel(bs.channel);
        if (chErr) errors.set('bank_switch.channel', chErr);
      }
      if (bs.cc !== undefined) {
        const ccErr = validators.cc(bs.cc);
        if (ccErr) errors.set('bank_switch.cc', ccErr);
      }
      if (bs.pc_base !== undefined) {
        const pcErr = validators.cc(bs.pc_base); // PC numbers use same range as CC
        if (pcErr) errors.set('bank_switch.pc_base', pcErr);
      }
      // Validate button numbers (1-indexed, must be within device button count)
      // STD10 can use buttons 1-10 or 11 (encoder push) for bank switching
      // Mini6 can use buttons 1-6 only (no encoder)
      const maxBankSwitchButton = config.device === 'mini6' ? 6 : 11;
      if (bs.button !== undefined) {
        if (bs.button < 1 || bs.button > maxBankSwitchButton) {
          errors.set('bank_switch.button', `Button must be between 1 and ${maxBankSwitchButton}`);
        }
      }
      if (bs.button_next !== undefined) {
        if (bs.button_next < 1 || bs.button_next > maxBankSwitchButton) {
          errors.set('bank_switch.button_next', `Button must be between 1 and ${maxBankSwitchButton}`);
        }
      }
      if (bs.button_prev !== undefined) {
        if (bs.button_prev < 1 || bs.button_prev > maxBankSwitchButton) {
          errors.set('bank_switch.button_prev', `Button must be between 1 and ${maxBankSwitchButton}`);
        }
      }
    }

    // Validate active_bank index
    if (config.active_bank !== undefined) {
      if (config.active_bank < 0 || config.active_bank >= config.banks.length) {
        errors.set('active_bank', `active_bank ${config.active_bank} out of range (max ${config.banks.length - 1})`);
      }
    }
  } else if (config.buttons) {
    // Legacy single-bank mode validation
    if (config.buttons.length !== expectedButtons) {
      errors.set('buttons', `Expected ${expectedButtons} buttons for ${config.device}, found ${config.buttons.length}`);
    }
    validateButtons(config.buttons, 'buttons', errors);
  }

  // Device-specific validation
  if (config.device === 'mini6') {
    if (config.encoder?.enabled) {
      errors.set('encoder.enabled', 'Mini6 does not support encoder');
    }
  }

  // Validate encoder
  if (config.encoder?.enabled) {
    const ccError = validators.cc(config.encoder.cc);
    if (ccError) errors.set('encoder.cc', ccError);

    if (config.encoder.channel !== undefined) {
      const chError = validators.channel(config.encoder.channel);
      if (chError) errors.set('encoder.channel', chError);
    }

    const min = config.encoder.min ?? 0;
    const max = config.encoder.max ?? 127;
    const rangeError = validators.range(min, max);
    if (rangeError) errors.set('encoder.range', rangeError);

    if (config.encoder.initial !== undefined) {
      const initError = validators.withinRange(config.encoder.initial, min, max);
      if (initError) errors.set('encoder.initial', `Initial ${initError.toLowerCase()}`);
    }

    if (config.encoder.push?.enabled) {
      const pushCcError = validators.cc(config.encoder.push.cc);
      if (pushCcError) errors.set('encoder.push.cc', pushCcError);

      if (config.encoder.push.channel !== undefined) {
        const chError = validators.channel(config.encoder.push.channel);
        if (chError) errors.set('encoder.push.channel', chError);
      }
      if (config.encoder.push.cc_on !== undefined) {
        const e = validators.cc(config.encoder.push.cc_on);
        if (e) errors.set('encoder.push.cc_on', e);
      }
      if (config.encoder.push.cc_off !== undefined) {
        const e = validators.cc(config.encoder.push.cc_off);
        if (e) errors.set('encoder.push.cc_off', e);
      }
    }
  }

  // Validate expression pedals
  for (const [key, exp] of [['exp1', config.expression?.exp1], ['exp2', config.expression?.exp2]] as const) {
    if (!exp?.enabled) continue;
    const p = `expression.${key}`;

    const ccError = validators.cc(exp.cc);
    if (ccError) errors.set(`${p}.cc`, ccError);

    if (exp.channel !== undefined) {
      const chError = validators.channel(exp.channel);
      if (chError) errors.set(`${p}.channel`, chError);
    }

    const min = exp.min ?? 0;
    const max = exp.max ?? 127;
    const rangeError = validators.range(min, max);
    if (rangeError) errors.set(`${p}.range`, rangeError);
  }

  return {
    isValid: errors.size === 0,
    errors,
  };
}

File: config-editor/src-tauri/src/config/validation.rs
//! Configuration validation logic
//!
//! Validates MIDI Captain configuration against device constraints and MIDI spec.

use super::models::*;
use super::types::*;

impl MidiCaptainConfig {
    /// Validate the configuration
    pub fn validate(&self) -> Result<(), Vec<String>> {
        let mut errors = Vec::new();

        // Validate global channel (0-15 internally, display as 1-16)
        if let Some(ch) = self.global_channel {
            if ch > 15 {
                errors.push(format!(
                    "Global channel value {} is invalid (must be 1-16, stored as 0-15)",
                    ch + 1
                ));
            }
        }

        // Check button count matches device
        let expected_buttons = match self.device {
            DeviceType::Std10 => 10,
            DeviceType::Mini6 => 6,
        };

        // Validate banks if present (multi-bank mode)
        if let Some(ref banks) = self.banks {
            if banks.is_empty() {
                errors.push("Banks array cannot be empty".to_string());
            } else if banks.len() > 8 {
                errors.push(format!("Max 8 banks supported, found {}", banks.len()));
            }

            for (bank_idx, bank) in banks.iter().enumerate() {
                let bank_num = bank_idx + 1;
                
                if bank.name.is_empty() {
                    errors.push(format!("Bank {} name cannot be empty", bank_num));
                } else if bank.name.len() > 20 {
                    errors.push(format!("Bank {} name '{}' exceeds 20 chars", bank_num, bank.name));
                }

                if bank.buttons.len() != expected_buttons {
                    errors.push(format!(
                        "Bank {} '{}': Expected {} buttons for {:?}, found {}",
                        bank_num, bank.name, expected_buttons, self.device, bank.buttons.len()
                    ));
                }

                self.validate_buttons(&bank.buttons, Some((bank_idx, &bank.name)), expected_buttons, &mut errors);
            }

            // Validate bank_switch config if present
            if let Some(ref bs) = self.bank_switch {
                if let Some(ch) = bs.channel {
                    if ch > 15 {
                        errors.push(format!("bank_switch.channel {} invalid (must be 0-15)", ch));
                    }
                }
                if let Some(cc) = bs.cc {
                    if cc > 127 {
                        errors.push(format!("bank_switch.cc {} exceeds 127", cc));
                    }
                }
                if let Some(pc_base) = bs.pc_base {
                    if pc_base > 127 {
                        errors.push(format!("bank_switch.pc_base {} exceeds 127", pc_base));
                    }
                }
                // Validate button numbers
                if let Some(btn) = bs.button {
                    if btn < 1 || btn > expected_buttons as u8 {
                        errors.push(format!("bank_switch.button {} out of range (1-{})", btn, expected_buttons));
                    }
                }
                if let Some(btn) = bs.button_next {
                    if btn < 1 || btn > expected_buttons as u8 {
                        errors.push(format!("bank_switch.button_next {} out of range (1-{})", btn, expected_buttons));
                    }
                }
                if let Some(btn) = bs.button_prev {
                    if btn < 1 || btn > expected_buttons as u8 {
                        errors.push(format!("bank_switch.button_prev {} out of range (1-{})", btn, expected_buttons));
                    }
                }
            }

            // Validate active_bank index
            if let Some(active_idx) = self.active_bank {
                if active_idx >= banks.len() as u8 {
                    errors.push(format!(
                        "active_bank {} out of range (max {})", 
                        active_idx, 
                        banks.len() - 1
                    ));
                }
            }
        }

        // Validate legacy buttons array if present (single-bank mode)
        // Skip legacy validation if banks are present (banks take precedence)
        if self.banks.is_none() {
            if let Some(ref buttons) = self.buttons {
                if buttons.len() != expected_buttons {
                    errors.push(format!(
                        "Expected {} buttons for {:?}, found {}",
                        expected_buttons,
                        self.device,
                        buttons.len()
                    ));
                }

                self.validate_buttons(buttons, None, expected_buttons, &mut errors);
            }
        } // end if self.banks.is_none()

        // Validate encoder if present
        if let Some(ref enc) = self.encoder {
            // Mini6 does not support encoder
            if self.device == DeviceType::Mini6 {
                errors.push("Mini6 does not support encoder".to_string());
            }
            if enc.cc > 127 {
                errors.push(format!("Encoder CC {} exceeds 127", enc.cc));
            }
            if enc.label.len() > 8 {
                errors.push(format!("Encoder label '{}' exceeds 8 chars", enc.label));
            }
            if enc.max < enc.min {
                errors.push(format!(
                    "Encoder max ({}) must be >= min ({})",
                    enc.max, enc.min
                ));
            }
            if enc.initial < enc.min || enc.initial > enc.max {
                errors.push(format!(
                    "Encoder initial ({}) must be between min ({}) and max ({})",
                    enc.initial, enc.min, enc.max
                ));
            }
            if let Some(ch) = enc.channel {
                if ch > 15 {
                    errors.push(format!(
                        "Encoder channel {} is invalid (must be 1-16)",
                        ch + 1
                    ));
                }
            }
            if let Some(ref push) = enc.push {
                if push.cc > 127 {
                    errors.push(format!("Encoder push CC {} exceeds 127", push.cc));
                }
                if push.label.len() > 8 {
                    errors.push(format!(
                        "Encoder push label '{}' exceeds 8 chars",
                        push.label
                    ));
                }
                if let Some(ch) = push.channel {
                    if ch > 15 {
                        errors.push(format!(
                            "Encoder push channel {} is invalid (must be 1-16)",
                            ch + 1
                        ));
                    }
                }
                if let Some(val) = push.cc_on {
                    if val > 127 {
                        errors.push(format!("Encoder push cc_on {} exceeds 127", val));
                    }
                }
                if let Some(val) = push.cc_off {
                    if val > 127 {
                        errors.push(format!("Encoder push cc_off {} exceeds 127", val));
                    }
                }
            }
        }

        // Validate expression pedals if present
        if let Some(ref exp) = self.expression {
            // Mini6 does not support expression pedals
            if self.device == DeviceType::Mini6 {
                errors.push("Mini6 does not support expression pedals".to_string());
            }
            if exp.exp1.cc > 127 {
                errors.push(format!("EXP1 CC {} exceeds 127", exp.exp1.cc));
            }
            if exp.exp1.label.len() > 8 {
                errors.push(format!("EXP1 label '{}' exceeds 8 chars", exp.exp1.label));
            }
            if exp.exp1.max < exp.exp1.min {
                errors.push(format!(
                    "EXP1 max ({}) must be >= min ({})",
                    exp.exp1.max, exp.exp1.min
                ));
            }
            if let Some(ch) = exp.exp1.channel {
                if ch > 15 {
                    errors.push(format!("EXP1 channel {} is invalid (must be 1-16)", ch + 1));
                }
            }
            if exp.exp2.cc > 127 {
                errors.push(format!("EXP2 CC {} exceeds 127", exp.exp2.cc));
            }
            if exp.exp2.label.len() > 8 {
                errors.push(format!("EXP2 label '{}' exceeds 8 chars", exp.exp2.label));
            }
            if exp.exp2.max < exp.exp2.min {
                errors.push(format!(
                    "EXP2 max ({}) must be >= min ({})",
                    exp.exp2.max, exp.exp2.min
                ));
            }
            if let Some(ch) = exp.exp2.channel {
                if ch > 15 {
                    errors.push(format!("EXP2 channel {} is invalid (must be 1-16)", ch + 1));
                }
            }
        }

        if errors.is_empty() {
            Ok(())
        } else {
            Err(errors)
        }
    }

    /// Validate an array of buttons (used for both legacy buttons and bank buttons)
    ///
    /// # Arguments
    /// * `buttons` - Array of button configurations to validate
    /// * `bank_info` - Optional (bank_index, bank_name) for error messages
    /// * `expected_count` - Expected number of buttons for the device
    /// * `errors` - Mutable vector to accumulate errors
    fn validate_buttons(
        &self,
        buttons: &[ButtonConfig],
        bank_info: Option<(usize, &str)>,
        expected_count: usize,
        errors: &mut Vec<String>,
    ) {
        let prefix = if let Some((idx, name)) = bank_info {
            format!("Bank {} '{}': ", idx + 1, name)
        } else {
            String::new()
        };

        let btn_path_prefix = if bank_info.is_some() {
            format!("banks[{}].buttons", bank_info.unwrap().0)
        } else {
            "buttons".to_string()
        };

        // Validate button count (already checked in caller, but keep for completeness)
        // Main validation loop
        for (i, button) in buttons.iter().enumerate() {
            let btn_num = i + 1;
            
            if let Some(cc) = button.cc {
                if cc > 127 {
                    errors.push(format!("{}Button {} CC {} exceeds 127", prefix, btn_num, cc));
                }
            }
            if button.label.len() > 6 {
                errors.push(format!(
                    "{}Button {} label '{}' exceeds 6 chars",
                    prefix, btn_num, button.label
                ));
            }
            if let Some(ref long_label) = button.long_press_label {
                if long_label.len() > 6 {
                    errors.push(format!(
                        "{}Button {} long_press_label '{}' exceeds 6 chars",
                        prefix, btn_num, long_label
                    ));
                }
            }
            if let Some(ch) = button.channel {
                if ch > 15 {
                    errors.push(format!(
                        "{}Button {} channel {} is invalid (must be 1-16)",
                        prefix, btn_num, ch + 1
                    ));
                }
            }
            if let Some(val) = button.cc_on {
                if val > 127 {
                    errors.push(format!("{}Button {} cc_on {} exceeds 127", prefix, btn_num, val));
                }
            }
            if let Some(val) = button.cc_off {
                if val > 127 {
                    errors.push(format!("{}Button {} cc_off {} exceeds 127", prefix, btn_num, val));
                }
            }
            if let Some(ms) = button.flash_ms {
                if ms < 50 || ms > 5000 {
                    errors.push(format!(
                        "{}Button {} flash_ms {} out of range (50-5000)",
                        prefix, btn_num, ms
                    ));
                }
            }
            if let Some(brightness) = button.dim_brightness {
                if brightness > 100 {
                    errors.push(format!(
                        "{}Button {} dim_brightness {} exceeds 100 (must be 0-100)",
                        prefix, btn_num, brightness
                    ));
                }
            }

            // Validate event command arrays (press, release, long_press, long_release)
            let validate_command = |cmd: &MidiCommand, event_name: &str, cmd_idx: usize| {
                let mut cmd_errors = Vec::new();
                if let Some(ch) = cmd.channel {
                    if ch > 15 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} channel {} invalid (must be 0-15)",
                            prefix, btn_num, event_name, cmd_idx, ch
                        ));
                    }
                }
                if let Some(cc) = cmd.cc {
                    if cc > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} cc {} exceeds 127",
                            prefix, btn_num, event_name, cmd_idx, cc
                        ));
                    }
                }
                if let Some(val) = cmd.value {
                    if val > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} value {} exceeds 127",
                            prefix, btn_num, event_name, cmd_idx, val
                        ));
                    }
                }
                if let Some(note) = cmd.note {
                    if note > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} note {} exceeds 127",
                            prefix, btn_num, event_name, cmd_idx, note
                        ));
                    }
                }
                if let Some(vel) = cmd.velocity {
                    if vel > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} velocity {} exceeds 127",
                            prefix, btn_num, event_name, cmd_idx, vel
                        ));
                    }
                }
                if let Some(prog) = cmd.program {
                    if prog > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} program {} exceeds 127",
                            prefix, btn_num, event_name, cmd_idx, prog
                        ));
                    }
                }
                if let Some(step) = cmd.pc_step {
                    if step < 1 || step > 127 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} pc_step {} out of range (1-127)",
                            prefix, btn_num, event_name, cmd_idx, step
                        ));
                    }
                }
                if let Some(thresh) = cmd.threshold_ms {
                    if thresh < 50 || thresh > 10000 {
                        cmd_errors.push(format!(
                            "{}Button {} {}.{} threshold_ms {} out of range (50-10000)",
                            prefix, btn_num, event_name, cmd_idx, thresh
                        ));
                    }
                }
                cmd_errors
            };

            if let Some(ref cmds) = button.press {
                for (idx, cmd) in cmds.iter().enumerate() {
                    errors.extend(validate_command(cmd, "press", idx));
                }
            }
            if let Some(ref cmds) = button.release {
                for (idx, cmd) in cmds.iter().enumerate() {
                    errors.extend(validate_command(cmd, "release", idx));
                }
            }
            if let Some(ref cmds) = button.long_press {
                for (idx, cmd) in cmds.iter().enumerate() {
                    errors.extend(validate_command(cmd, "long_press", idx));
                }
            }
            if let Some(ref cmds) = button.long_release {
                for (idx, cmd) in cmds.iter().enumerate() {
                    errors.extend(validate_command(cmd, "long_release", idx));
                }
            }

            // Validate per-state event arrays (states[*].press/release/long_press/long_release)
            if let Some(ref states) = button.states {
                for (state_idx, state) in states.iter().enumerate() {
                    let state_num = state_idx + 1;

                    if let Some(ref cmds) = state.press {
                        for (cmd_idx, cmd) in cmds.iter().enumerate() {
                            errors.extend(validate_command(
                                cmd,
                                &format!("states[{}].press", state_num),
                                cmd_idx,
                            ));
                        }
                    }
                    if let Some(ref cmds) = state.release {
                        for (cmd_idx, cmd) in cmds.iter().enumerate() {
                            errors.extend(validate_command(
                                cmd,
                                &format!("states[{}].release", state_num),
                                cmd_idx,
                            ));
                        }
                    }
                    if let Some(ref cmds) = state.long_press {
                        for (cmd_idx, cmd) in cmds.iter().enumerate() {
                            errors.extend(validate_command(
                                cmd,
                                &format!("states[{}].long_press", state_num),
                                cmd_idx,
                            ));
                        }
                    }
                    if let Some(ref cmds) = state.long_release {
                        for (cmd_idx, cmd) in cmds.iter().enumerate() {
                            errors.extend(validate_command(
                                cmd,
                                &format!("states[{}].long_release", state_num),
                                cmd_idx,
                            ));
                        }
                    }
                }
            }

            // select_group rules: not allowed with momentary, tap, or keytimes > 1
            if let Some(_) = button.select_group {
                if button.mode == ButtonMode::Momentary {
                    errors.push(format!(
                        "{}Button {} select_group not supported for momentary mode",
                        prefix, btn_num
                    ));
                }
                if button.mode == ButtonMode::Tap {
                    errors.push(format!(
                        "{}Button {} select_group not supported for tap mode",
                        prefix, btn_num
                    ));
                }
                if let Some(kt) = button.keytimes {
                    if kt > 1 {
                        errors.push(format!(
                            "{}Button {} select_group not supported with keytimes > 1",
                            prefix, btn_num
                        ));
                    }
                }
            }
            if let Some(ds) = button.default_selected {
                // no extra check here; normalization performed elsewhere
                let _ = ds;
            }
        }
    }
}
